### PR TITLE
Narrow errors for neo datetime

### PR DESCRIPTION
**Pull request by robertbastian**
Part of #4336

Per policy in https://github.com/unicode-org/icu4x/pull/4638#issuecomment-1994218244

### CODE REVIEW

File: components/datetime/src/external_loaders.rs [hunk 1]
```diff
@@ -75,24 +75,33 @@ impl FixedDecimalFormatterLoader for ExternalLoaderCompiledData {
         &self,
         locale: &DataLocale,
         options: FixedDecimalFormatterOptions,
-    ) -> Result<FixedDecimalFormatter, DecimalError> {
-        FixedDecimalFormatter::try_new(locale, options)
+    ) -> Result<FixedDecimalFormatter, DataError> {
+        FixedDecimalFormatter::try_new(locale, options).map_err(|e| match e {
+            DecimalError::Data(e) => e,
+            _ => unreachable!(),
```
Comment (robertbastian) [inline components/datetime/src/external_loaders.rs:81]: this unreachable branch will disappear in 2.0, when the external constructor can return `DataError` only

**Comment by sffc** [comment]
@robertbastian Can you reference the corresponding issue in the OP?

File: components/datetime/src/format/neo.rs [hunk 2]
```diff
@@ -952,7 +994,7 @@ impl RawDateTimeNames {
                 ///// Textual symbols /////
                 FieldSymbol::Era => {
                     self.load_year_names(
-                        year_provider.ok_or(Error::MissingNames(field))?,
+                        year_provider.ok_or(LoadError::MissingNames(field))?,
```
Comment (sffc) [inline components/datetime/src/format/neo.rs:997]: Thought: With the work I'm doing to consolidate the neo formatter types, this error case _might_ go away which would be nice.

**Comment by robertbastian** [comment]
Done

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
@robertbastian Also please link the issue about constructor errors, where we listed out the specific cases where returning DataError directly from constructors was acceptable

**Pull request merged**